### DISCUSSION TITLE
Shaders workflow

### DISCUSSION THREAD

**Discussion by rudiHammad** [axmolengine/axmol#2261] [Q&A] [answered]
Hi,
I wanted to learn a bit more about shaders and I found this great doc by @halx99 and @rh101 https://github.com/axmolengine/axmol/wiki/Shaders-in-Axmol
I am just having some trouble to understand how to proceed. For instance it says:

> To use your own custom shaders in your project, you simply need to place the shaders into {project_root}/Source/shaders/.
Once you build your project, the shaders will processed and placed into ${CMAKE_BINARY_DIR}/runtime/axslc/custom. The entire contents of the ${CMAKE_BINARY_DIR}/runtime/axslc/ folder are copied to app_res_root/axslc.
In code, you reference the custom shaders by prefixing the filename with custom/.
For example, if your shaders are {project_root}/Source/shaders/MyEffect.fsh and {project_root}/Source/shaders/MyEffect.vsh, then the shaders will be processed and saved into ${CMAKE_BINARY_DIR}/runtime/axslc/custom as MyEffect_fs and MyEffect_vs.

But in my case I build my project with any shaders. I am using axmol in -2.1.3.
So how would I proceed from here. Should I just directly write the shaders in ${CMAKE_BINARY_DIR}/runtime/axslc/custom as xxx_vs and xxx_fs?

I also have a question about the path. Just to fort testing I copied and renamed existing working shaders as foo_vs and foo_fs.
And then I tried the code bellow which failed.

` auto* program = ProgramManager::getInstance()->loadProgram("shaders/foo_vs", "shaders/foo_fs");`

So my guess is that, as I mentioned, the workflow to create new shaders is to write them directly in ${CMAKE_BINARY_DIR}/runtime/axslc/custom as xxx_vs and xxx_fs?

thanks

**Comment by rh101** [comment, marked as answer]
The guide is for Axmol v2, so still follow the guide for Axmol v2.1.3.

> Should I just directly write the shaders in ${CMAKE_BINARY_DIR}/runtime/axslc/custom as xxx_vs and xxx_fs?

No, you still put the files in your `{project_dir}/Source/shaders/`, as "myshader.frag" etc. etc.

They will be compiled at build time, and automatically placed in the correct folder.

So, for "myshader.frag", it would be referenced as `custom/myshader_fs` in your code.

Just follow the guide.  Do not just copy files into the build folders as you've done.

The guide has been updated to remove a section that may have been the source of confusion.

**Reply by rudiHammad** [reply]
Cool, thanks. Now it's clear.
By the way, if the game we build is for desktop and not mobiles, should we use GLSL  instead of ESSL?
That's what I read doing a google quick search.

**Reply by halx99** [reply]
ANGLE should be ESSL
